cmd/internal/kube/daemonset/daemonset.go

```diff
@@ -17,40 +19,115 @@ import (
 	"github.com/postmanlabs/postman-insights-agent/printer"
 	"github.com/postmanlabs/postman-insights-agent/rest"
 	"github.com/postmanlabs/postman-insights-agent/telemetry"
+	coreV1 "k8s.io/api/core/v1"
 )
 
 const (
 	apiContextTimeout = 20 * time.Second
 )
 
+type PodTrafficMonitorStage int
+
+// These are different stages of pod traffic monitoring
+// PodDetected/PodInitialized -> TrafficMonitoringStarted -> TrafficMonitoringFailed/TrafficMonitoringEnded/PodTerminated -> TrafficMonitoringStopped -> PodRemovedFromMap
+const (
+	// When agent finds an already running pod
+	PodDetected PodTrafficMonitorStage = iota
+
+	// When agent will receive pod created event
+	PodInitialized
+
+	// When apidump process is started for the pod
+	TrafficMonitoringStarted
+
+	// When apidump process is errored for the pod
+	TrafficMonitoringFailed
+
+	// When apidump process is ended without any issue for the pod
+	TrafficMonitoringEnded
+
+	// When agent will receive pod deleted event
+	PodTerminated
+
+	// When apidump process is stopped for the pod
+	TrafficMonitoringStopped
+
+	// Final stage after which pod will be removed from the map
+	PodRemovedFromMap
+)
+
+type PodCreds struct {
+	InsightsAPIKey      string
+	InsightsEnvironment string
+}
+
 type PodArgs struct {
 	// apidump related fields
 	InsightsProjectID        akid.ServiceID
-	InsightsAPIKey           string
 	InsightsReproModeEnabled bool
 
 	// Pod related fields
 	PodName       string
 	ContainerUUID string
+	PodCreds      PodCreds
+
+	// for state management
+	PodTrafficMonitorStage PodTrafficMonitorStage
+	StageChangeMutex       *sync.Mutex
+
+	// send stop signal to apidump process
+	StopChan chan error
+}
+
+func (p *PodArgs) setPodTrafficMonitorStage(stage PodTrafficMonitorStage) {
+	p.StageChangeMutex.Lock()
+	defer p.StageChangeMutex.Unlock()
+	p.PodTrafficMonitorStage = stage
+}
+
+func (p *PodArgs) validatePodTrafficMonitorStage(
+	nextStage PodTrafficMonitorStage,
+	allowedPriorStages ...PodTrafficMonitorStage,
+) (bool, error) {
+	if slices.Contains(allowedPriorStages, p.PodTrafficMonitorStage) {
+		return false, nil
+	}
+
+	if p.PodTrafficMonitorStage == nextStage {
+		printer.Debugf("API dump process for pod %s is already in state %d", p.PodName, nextStage)
```

Comment:
Is this low-volume enough that we should emit it at a higher level for production monitoring?